### PR TITLE
Default interface members prog guide

### LINKED ISSUE #12489: Language Reference, Programming guide: Default interface members updates
It's likely that almost all mentions of interfaces in the programming guide and the language reference make some passing comment that interfaces cannot contain implementation. That's changing

### LINKED ISSUE #14803: No mention of the new access modifiers in interfaces in C 8.0
The 8.0 proposal is talking about the `private` and `internal` modifiers in  interfaces in 8.0 but no mention of them in what's new page. This is a big change, which I believe should have been mentioned in this page.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: a94537c5-754e-9dea-94a8-7573011f588c
* Version Independent ID: aba09118-22c5-2c6a-954a-b9659a9644ae
* Content: [What's new in C# 8.0 - C# Guide](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#feedback)
* Content Source: [docs/csharp/whats-new/csharp-8.md](https://github.com/dotnet/docs/blob/master/docs/csharp/whats-new/csharp-8.md)
* Product: **dotnet-csharp**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### LINKED ISSUE #16821: Need clarification whether inheriting from child interface implies inheriting from parent interface
Consider
```
interface Parent { }
interface Child : Parent { }
class Class : Child { }
```
Is `(Parent) new Class()` legal?  The current version of the article does not clearly indicate whether `Class` automatically inherits `Parent`, or in other words, whether the class definition is equivalent to `class Class : Parent, Child { }`.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 3e982a51-097f-688f-710c-1dfcf313da5e
* Version Independent ID: 8691f826-62cb-da0a-8460-a38aa2bc2a70
* Content: [Interfaces - C# Programming Guide](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/interfaces/)
* Content Source: [docs/csharp/programming-guide/interfaces/index.md](https://github.com/dotnet/docs/blob/master/docs/csharp/programming-guide/interfaces/index.md)
* Product: **dotnet-csharp**
* Technology: **csharp-fundamentals**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #14803 
Fixes #12489 
Fixes #16821

Depends on dotnet/samples#1998

Review all occurrences of "interface" in the C# programming guide. Update any that state or imply that interface members cannot contain implementation. They can, starting in C# 8.0

While working on those files, do general cleanup:

- check that all snippets are in a project that compiles.
- Various formatting updates.
- Run acrolinx.

Reviews will be easier with hide whitespace.

### CODE REVIEW

**Comment by pkulikov** [comment]
@BillWagner do you think the [Methods](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/methods) article should be updated as well? Now it claims:

> Methods are declared in a class or struct by specifying ...

implying that one cannot define a method in an interface.

**Comment by BillWagner** [comment]
@pkulikov 

> do you think the Methods article should be updated as well?

Yes. I'll open a new PR for that fix.

**Comment by BillWagner** [comment]
closing and reopening for a fresh build

**Comment by BillWagner** [comment]
switching reviewers to @Thraka

File: docs/csharp/programming-guide/classes-and-structs/polymorphism.md [hunk 1]
```diff
@@ -1,93 +1,108 @@
...
 > [!NOTE]
-> It is recommended that virtual members use `base` to call the base class implementation of that member in their own implementation. Letting the base class behavior occur enables the derived class to concentrate on implementing behavior specific to the derived class. If the base class implementation is not called, it is up to the derived class to make their behavior compatible with the behavior of the base class.  
-  
-## In This Section  
-  
-- [Versioning with the Override and New Keywords](./versioning-with-the-override-and-new-keywords.md)  
-  
-- [Knowing When to Use Override and New Keywords](./knowing-when-to-use-override-and-new-keywords.md)  
-  
+> It is recommended that virtual members use `base` to call the base class implementation of that member in their own implementation. Letting the base class behavior occur enables the derived class to concentrate on implementing behavior specific to the derived class. If the base class implementation is not called, it is up to the derived class to make their behavior compatible with the behavior of the base class.
```
Comment (Thraka) [inline docs/csharp/programming-guide/classes-and-structs/polymorphism.md]: Is there some prescriptive guidance on this we can link to? For example, is there a convention around naming a virtual/abstract member `OnDoSomething` you're indicating a callback for the purpose of being overridden and therefor you don't need to call `base.OnDoSomething`? I've always wondered that. I did a quick search but didn't find anything. Maybe you know..

**Review by Thraka — CHANGES_REQUESTED** [review]
Few things to fix. And snippets not showing yet.

File: docs/csharp/programming-guide/concepts/object-oriented-programming.md [hunk 2]
```diff
@@ -116,22 +74,22 @@ class SampleClass
     public int Sample
     {
         // Return the value stored in a field.
-        get { return _sample; }
+        get => _sample;
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/programming-guide/concepts/object-oriented-programming.md]: The first sample above shows an auto-implemented property. The text above this paragraph explains that this second sample isn't auto-implemented.

File: docs/csharp/programming-guide/classes-and-structs/polymorphism.md (hunk 1 above)
Comment (BillWagner) [reply docs/csharp/programming-guide/classes-and-structs/polymorphism.md]: Andy, I don't know of any prescriptive guidance here.

**Comment by BillWagner** [comment]
closing and reopening for a new build.

**Comment by BillWagner** [comment]
Thanks @Thraka 

I've addressed all the feedback.

File: docs/csharp/programming-guide/concepts/object-oriented-programming.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Thraka) [reply docs/csharp/programming-guide/concepts/object-oriented-programming.md]: Oh I see, I missed the fact that the text above was collapsed and I was reading the wrong intro.

**Review by Thraka — APPROVED** [review]

**Pull request merged**